packages/client/hmi-client/src/services/notebook.ts

```diff
@@ -0,0 +1,25 @@
+import { cloneDeep } from 'lodash';
+import { WorkflowNode } from '@/types/workflow';
+
+export interface NotebookHistory {
+	code: string;
+	timestamp: number;
+}
+
+// A common pattern used to save code from a notebook within an operator
+// This is ready to be ported to nodes such as tera-model-edit, tera-model-config and tera-stratify-mira
+// Not ported yet since this will ruin the states of the nodes that already exist due to their differently named properties
+export const saveCodeToState = (node: WorkflowNode<any>, code: string, hasCodeRun: boolean) => {
+	const state = cloneDeep(node.state);
+
+	state.hasCodeRun = hasCodeRun;
+	// for now only save the last code executed, may want to save all code executed in the future
+	const notebookHistoryLength = node.state.notebookHistory.length;
+	const timestamp = Date.now();
+	if (notebookHistoryLength > 0) {
+		state.notebookHistory[0] = { code, timestamp };
+	} else {
+		state.notebookHistory.push({ code, timestamp });
+	}
+	return state;
+};
```

Comment:
You may need to do a few sanity checks and not assume that `node.state` has all the fields that you are using.